### ISSUE TITLE
Scrobbles last played track twice.

### ISSUE THREAD

**Issue by augustonasser** [tgwizard/sls#80]
After I play an album, the last track is scrobbled but is kept on cache. When I play something again, SLS scrobbles this cached scrobble, so the song is scrobbled twice.

Galaxy X (brazilian Galaxy Nexus).
Default music player (Google Play).

**Comment by Sagehorn** [comment]
Same problem with google play music and shuttle. SLS scrobbles twice when single/last(?) track is played..

**Comment by HumbleBeeBumbleBee** [comment]
Duplicate #52
Potential solution. https://github.com/tgwizard/sls/pull/149 Check it out!

**Comment by HumbleBeeBumbleBee** [comment]
@augustonasser @timurrrr @kevjs1982 @Voodooguru @JohnMaguire @barsnick @arkdae @Sagehorn 
If you know how download and build the .apk
Please use this branch https://github.com/tgwizard/sls/tree/testing-149
Otherwise you can use the latest .apk here.
https://drive.google.com/file/d/0B9lPaMaVVtiJLVlLendjSmxRZDA/view

Please tell me if this fixed your issues. Thanks!

**Comment by arkdae** [comment]
This not only doesn't fix the issue for me, but seems to make it worse.

Firstly, I am not confident that I have this APK. When I go to about, it still lists as version 1.48. A build number or some other way to verify that I have installed this testing version would be helpful. I did uninstall the one from Google Play, because it said this one has a different signature, and then I clicked on the link in the email to install this one (after enabling unknown sources on my phone).

The original issue for me is two fold. First is as reported here. After completing a queue, if you start a new queue, the last song from the last queue will be scrobbled a second time. I often select whole albums to play, not just individual songs. So, in my example, I can select an album and hit the orange play button. When it is done, I can select a new album, hit the orange play button, and the last song from the last album will be scrobbled a second time as it starts playing the new album.

The second issue has been touched upon in this thread. I often pause and play music throughout the day, and from one day to the next. If I pause the music at night, for example, and then start it up again the next morning, I will have two scrobbles for the song, even though I never restarted the song, just picked up where it left off.

Now, with this version, I find that the second issue is worse. Previously, I would only get duplicate scrobbles. Now, I am getting many repeated scrobbles. It has only happened twice in the two days I have been using this version, but I have gotten six and seven identical consecutive scrobbles from this issue, where I would normally get only two.

So, please, undo this change, as it makes it worse for me.

Additional info: I'm running this on a Sony Xperia Z3V, with OS version 4.4.4

**Comment by arkdae** [comment]
Also, I would say that this is not a duplicate of issue #52 . In that, he states that the issue happens after every song, not just the last in a queue. Also, he states that it was fixed when he changed hardware.

**Comment by arkdae** [comment]
Sorry for the repeated messages, but one more thing. What is the option "Also scrobble when you've finished listening to a playlist?" I can't imagine what this will do, except scrobble a second time. I've disabled it, but it does not help.

**Comment by HumbleBeeBumbleBee** [comment]
@arkdae  THANK YOU!!! Your feedback is much appreciated.
I am in the process of reverting a bunch of changes.
Please, send as many messages as you like.
I am currently having a shortage of testers.
Updates are useless if nobody tests them before the next release.

**Comment by metanota** [comment]
@arkdae You haven't named the app that you're using for listening and scrobbling music so we can only guess, why it happens to be much harder for you then for anyone else ;)

There are my thoughts about the situation: your app sends both generic android intents together with app's specific ones. (I know for sure that PowerAmp does that, for example). Every intent processes separately, while poor sls tracks only one context at a time. It handles these different intents as like the user turn's on some another app for playing music, and decides that the track needs to be sent to server for every switching player (if it may be scrobbled at this point, of course). This behaviour is legit since user may be really switching his app... but it ruins everything when app is designed to send different intents from several names on every action.

When the application sends generic android intents, there is "Android Music Player" point appears in the "Enabled apps" menu point. It says "Won't scrobble music from this app"; do not believe it! For now I can't fix this programmatically, so users should uncheck this point manually. This was the issue before #149 and will remain after. sadly. @arkdae you may probably already done it before; but since you've completely removed the app to switch it with another one (with the different signature) the settings were restored to default "all checked" value.

**TL;DR**
@arkdae: go to the "Enabled apps" preference an uncheck all the scrobbling applications that you aren't using. "Android Music Player" is the most suspicious point in the list. Write me if it won't help.

developers: this issue have bothered me for the long time (see #94, #119, #123 for example) and I still didn't found the right solution. Forcing users to do some extra actions in preferences is very bad design decision (especially when the relation is unclear and punishes for not knowing it at the same time).

**Comment by arkdae** [comment]
No need to guess, just ask nicely. I am using Google Play Music. I did mention that in my first comment on this thread, but forgot to repeat it. For that I am deeply sorry.

**Comment by arkdae** [comment]
The only application listed in "Enabled apps" is "Android Music Player" I have unchecked that and I will monitor for any changes in behavior.

**Comment by arkdae** [comment]
When I uncheck "Android Music Player," there are no longer any scrobbles from Google Play Music to last.fm.

**Comment by arkdae** [comment]
This is interesting. After I completed playing an album, I then checked "Android Music Player" to resume scrobbling again. I selected a new album, and the last track of the previous album was scrobbled. My guess is that Google Play Music is sending some message to scrobble that last track again when you start playing a new song.

Perhaps there is some way to inspect the message or "intent" as you put it, to differentiate these secondary messages from actual plays?

Or, could some timing be put into place? My observation with duplicate scrobbles of the last track of a queue is that it doesn't scrobble until you start a new album or song. So, perhaps you'll get two messages in rapid succession, the last song, and the new one. If you just drop the first message, that should deal with duplicates in this instance. Since songs usually last more than one second, messages in rapid succession do not indicate actual plays.

**Comment by HumbleBeeBumbleBee** [comment]
Hello. If anybody wants a copy of the newest SLS that has fixed some bugs with Google Play Music.
Please let me know. I provide support and maintenance for the lead developers.

Here is the main issue with scrobbling.


There are two major issues with doing this and two major compatibility issues.

SLS provides support for a variety of applications.
OR
SLS provides support for single track repeat and more structured scrobbling.

Some of these applications don't share track duration (well).
and
Some of these applications don't tell SLS when the track is paused or completed, so SLS cannot find out when the track officially started or finished.

I am manually going through all the applications to see which ones don't work with a more compatible scrobbling service.

The Queue method works for now.

**Comment by HumbleBeeBumbleBee** [comment]
Please try this version.

https://drive.google.com/open?id=0B9lPaMaVVtiJaGwwdkt0UjBNVHc
